File: frontend/lib/features/marketplace/data/models/order_model.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import '../../domain/entities/order.dart';

class OrderModel extends OrderEntity {
  const OrderModel({
    required super.id,
    required super.productId,
    required super.productName,
    required super.buyerId,
    required super.buyerName,
    required super.sellerId,
    required super.sellerName,
    required super.quantity,
    required super.unit,
    required super.pricePerUnit,
    required super.totalPrice,
    required super.status,
    super.notes,
    required super.location,
    required super.createdAt,
    required super.updatedAt,
  });

  factory OrderModel.fromFirestore(DocumentSnapshot doc) {
    final d = doc.data() as Map<String, dynamic>;
    return OrderModel(
      id: doc.id,
      productId: d['productId'] ?? '',
      productName: d['productName'] ?? '',
      buyerId: d['buyerId'] ?? '',
      buyerName: d['buyerName'] ?? '',
      sellerId: d['sellerId'] ?? '',
      sellerName: d['sellerName'] ?? '',
      quantity: (d['quantity'] as num?)?.toDouble() ?? 0,
      unit: d['unit'] ?? 'kg',
      pricePerUnit: (d['pricePerUnit'] as num?)?.toDouble() ?? 0,
      totalPrice: (d['totalPrice'] as num?)?.toDouble() ?? 0,
      status: d['status'] ?? 'pending',
      notes: d['notes'],
      location: d['location'] ?? '',
      createdAt: _ts(d['createdAt']),
      updatedAt: _ts(d['updatedAt']),
    );
  }

  factory OrderModel.fromJson(Map<String, dynamic> d) {
    return OrderModel(
      id: d['id'] as String? ?? '',
      productId: d['productId'] as String? ?? '',
      productName: d['productName'] as String? ?? '',
      buyerId: d['buyerId'] as String? ?? '',
      buyerName: d['buyerName'] as String? ?? '',
      sellerId: d['sellerId'] as String? ?? '',
      sellerName: d['sellerName'] as String? ?? '',
      quantity: (d['quantity'] as num?)?.toDouble() ?? 0,
      unit: d['unit'] as String? ?? 'kg',
      pricePerUnit: (d['pricePerUnit'] as num?)?.toDouble() ?? 0,
      totalPrice: (d['totalPrice'] as num?)?.toDouble() ?? 0,
      status: d['status'] as String? ?? 'pending',
      notes: d['notes'] as String?,
      location: d['location'] as String? ?? '',
      createdAt: _parseDate(d['createdAt']),
      updatedAt: _parseDate(d['updatedAt']),
    );
  }

  factory OrderModel.fromEntity(OrderEntity e) => OrderModel(
        id: e.id,
        productId: e.productId,
        productName: e.productName,
        buyerId: e.buyerId,
        buyerName: e.buyerName,
        sellerId: e.sellerId,
        sellerName: e.sellerName,
        quantity: e.quantity,
        unit: e.unit,
        pricePerUnit: e.pricePerUnit,
        totalPrice: e.totalPrice,
        status: e.status,
        notes: e.notes,
        location: e.location,
        createdAt: e.createdAt,
        updatedAt: e.updatedAt,
      );

  Map<String, dynamic> toFirestore() => {
        'productId': productId,
        'productName': productName,
        'buyerId': buyerId,
        'buyerName': buyerName,
        'sellerId': sellerId,
        'sellerName': sellerName,
        'quantity': quantity,
        'unit': unit,
        'pricePerUnit': pricePerUnit,
        'totalPrice': totalPrice,
        'status': status,
        'notes': notes,
        'location': location,
        'createdAt': Timestamp.fromDate(createdAt),
        'updatedAt': Timestamp.fromDate(updatedAt),
      };

  static DateTime _ts(dynamic v) {
    if (v is Timestamp) return v.toDate();
    if (v is DateTime) return v;
    return DateTime.now();
  }

  static DateTime _parseDate(dynamic v) {
    if (v is Timestamp) return v.toDate();
    if (v is DateTime) return v;
    if (v is String && v.isNotEmpty) {
      return DateTime.tryParse(v) ?? DateTime.now();
    }
    return DateTime.now();
  }
}

File: backend/models/order_model.py
# backend/models/order_model.py
from dataclasses import dataclass
from typing import Optional
from utils.helpers import ts_to_iso

@dataclass
class OrderModel:
    id:            str
    product_id:    str
    product_name:  str
    buyer_id:      str
    buyer_name:    str
    seller_id:     str
    seller_name:   str
    quantity:      float
    unit:          str
    price_per_unit: float
    total_price:   float
    status:        str        # pending | accepted | rejected | delivered
    location:      str
    created_at:    str
    updated_at:    str
    notes:         Optional[str] = None

    def to_dict(self) -> dict:
        return {
            "id":           self.id,
            "productId":    self.product_id,
            "productName":  self.product_name,
            "buyerId":      self.buyer_id,
            "buyerName":    self.buyer_name,
            "sellerId":     self.seller_id,
            "sellerName":   self.seller_name,
            "quantity":     self.quantity,
            "unit":         self.unit,
            "pricePerUnit": self.price_per_unit,
            "totalPrice":   self.total_price,
            "status":       self.status,
            "notes":        self.notes,
            "location":     self.location,
            "createdAt":    self.created_at,
            "updatedAt":    self.updated_at,
        }

    @classmethod
    def from_firestore(cls, doc_id: str, data: dict) -> "OrderModel":
        return cls(
            id=             doc_id,
            product_id=     data.get("productId",   ""),
            product_name=   data.get("productName", ""),
            buyer_id=       data.get("buyerId",     ""),
            buyer_name=     data.get("buyerName",   ""),
            seller_id=      data.get("sellerId",    ""),
            seller_name=    data.get("sellerName",  ""),
            quantity=       float(data.get("quantity",     0)),
            unit=           data.get("unit", "kg"),
            price_per_unit= float(data.get("pricePerUnit", 0)),
            total_price=    float(data.get("totalPrice",   0)),
            status=         data.get("status", "pending"),
            notes=          data.get("notes"),
            location=       data.get("location", ""),
            created_at=     ts_to_iso(data.get("createdAt")),
            updated_at=     ts_to_iso(data.get("updatedAt")),
        )
